### ISSUE TITLE
Graph transformations could be a big thing

### ISSUE THREAD

**Issue by oxinabox** [invenia/TimeDag.jl#5]
I think a lot of cool things could be expressed as a graph transformation that is run on the graph to get a new graph that is executed.

For example one could add inter-execute call caching by replacing all nodes (or a subset filtered with metadata) with a new cached node that checks on disk for a cache of it, and if it finds it, returns it, and if not generates and saves it.

More complex:
To speed up database queries for e.g. rolling windows,
you could walk the graph to find all "fetch" nodes.
Union those into some set of prefetch nodes that do a big query to make a local datastore, with is then fetched against by each operation.
(This can also be expressed without doing this though)

**Comment by tpgillam** [comment]
Yep, I definitely agree there's a lot of cool things we could do :)

Personally I'm most excited by transformations that will give _provably identical_ results, just faster or more efficiently. I think a lot of these would boil down to some form of pattern matching. (The one gotcha is making sure that alignments are treated correctly, but this is definitely soluble.)

Even better, there's a subset of these that can be applied whilst the graph is first being constructed, without rewriting. Some trivial ones:
```
lag(n, 0) === n
-(-n) === n
```
i.e. whenever we can prove that we already have a node in the graph that is identical to the one we're trying to compute, we re-use it.
